org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,781 @@
+
+        // In the event we haven't explicitly set cert and key, try and infer
+        // them from the alias specified... We assume that when the SSLEngine
+        // has a certificate already, we want to use them, even if parsed has
+        // a null certificate.
+        if (parsed.getAlias() != null && key_managers != null && key_managers.length > 0 && cert == null && key == null) {
+            setCertFromAlias(parsed.getAlias());
+        }
+
+        // When we have a value for the peer hostname, we should try and use
+        // it.
+        if (parsed.getHostname() != null) {
+            setHostname(parsed.getHostname());
+        }
+    }
+
+    /**
+     * Set the hostname used to validate the peer's certificate.
+     *
+     * This is usually passed to NSS via a call to NSS's ill-named
+     * SSL_SetURL(...), which really takes a hostname.
+     *
+     * Note: if this isn't called (and no peerHost was specified via a
+     * constructor), NSS will accept any host name provided by the server!
+     * Only useful for validating the server certificate; not used when
+     * validating the peer's certificate.
+     */
+    public void setHostname(String name) {
+        // Note, this is the only way to set the hostname used for validation.
+        //
+        // In the event setHostname is never explicitly called, we should try
+        // and fall back to the value passed in the constructor.
+        hostname = name;
+    }
+
+    /**
+     * Choose a certificate to give to the peer from the specified alias,
+     * assuming KeyManagers have already been specified and at least one is
+     * a JSSKeyManager.
+     *
+     * When alias is null, this clears all previous certs and keys.
+     *
+     * If no KeyManagers have been specified, raises an
+     * IllegalArgumentException stating as much.
+     *
+     */
+    public void setCertFromAlias(String alias) throws IllegalArgumentException {
+        if (alias == null) {
+            // Per calling, semantics, get rid of any existing cert/key we
+            // might have.
+            cert = null;
+            key = null;
+            return;
+        }
+
+        if (key_managers == null || key_managers.length == 0) {
+            String msg = "Missing or null KeyManagers; refusing to search ";
+            msg += "for cert";
+            throw new IllegalArgumentException(msg);
+        }
+
+        for (X509KeyManager key_manager : key_managers) {
+            if (key_manager == null) {
+                // Skip this key_manager. This case could occur when
+                // setKeyManagers(...) is passed an array containing the value
+                // null, but otherwise shouldn't happen.
+                continue;
+            }
+
+            if (!(key_manager instanceof JSSKeyManager)) {
+                // We're explicitly looking for a JSSKeyManager; skip this if
+                // it doesn't match.
+                continue;
+            }
+
```

Comment:
Currently there is only one `JSSKeyManager`, and it can't be configured in any different ways. That means its functional equivalent so far. We could probably add a break if we detect we have a non-`null` cert and key.